### ISSUE TITLE
Refresh UE? autocomplete widgets, after using rgthree pan / zoom nodes

### ISSUE THREAD

**Issue by JorgeR81** [chrisgoringe/cg-use-everywhere#173]
The rgthree suite has some nodes that allow to pan / zoom the workflow.

There is the Fast Groups Muter, that also can take you the group's location in the workflow.
And there is also a "bookmark" node, which takes us to it's position, with a pre-selected zoom level, via keybind. 
https://github.com/rgthree/rgthree-comfy?tab=readme-ov-file#bookmark-

When we use them, the UE? autocomplete widgets don't refresh fast enough.
So we can still see them, in the position and size that they had before.

And we can also click on them, like a regular widget, to show the text input popup, and the autocomplete balloon.
So it's not just a visual artifact. 

After panning manually a bit, this goes away. 
But the point of using these rgthree nodes, is to get to that exact place in the workflow automatically.
So having to pan manually afterwards, makes them far less useful.

![rg3o](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/66f66e6a-3263-4547-8128-4231bf912221)

**Comment by chrisgoringe** [comment]
Interesting. Does it happen with any other modes? In particular, anything with a multiline text input would be interesting to test.

**Comment by JorgeR81** [comment]
This is the only issue I noticed using the rgthree nodes. 
The large field for the prompt text does not have this issue. 

What I noticed, sometimes, is that when I'm making changes in more complex workflows, the autocomplete widgets still appear in the top left corner, like before.
There are multiple ones, on top of each other. 
And then, they also disappear after planning / zooming, like before. 
But I'm not entirely sure about what actions trigger that. 
One thing I noticed is that using Ctrl+Z ( Undo ) can trigger that.

**Comment by JorgeR81** [comment]
> Does it happen with any other modes? In particular, anything with a multiline text input would be interesting to test.

In a complex workflow, the main issue can also happen, even without using the rg3 zoom / pan nodes.

Here I have a workflow with "priory boost" up to 5.
So there is a lot of text in the EU? widgets.

When I zoom in manually, I can still see the widgets in the screen's corner, from a previous zoom level.  

![b5](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/32b91cb8-b70b-4da4-b784-8589b904c352)

**Comment by JorgeR81** [comment]
On a side note, 
And sorry for bringing up up this again, 

But, are you sure that, in the long term, wouldn't it just be easier to have a custom autocomplete, instead of trying to hunt down all these widget errors ?

ComfyUI-Custom-Scripts uses this MIT implementation, and I never noticed any issues.

https://github.com/pythongosssss/ComfyUI-Custom-Scripts?tab=readme-ov-file#autocomplete

https://github.com/pythongosssss/ComfyUI-Custom-Scripts/blob/main/web/js/common/autocomplete.js

**Comment by JorgeR81** [comment]
OK, the current autocomplete widget does have an advantage.

It can display a larger number of characters.
 
All single line string nodes, from different suites, seem to have the same limit on the characters displayed. 
Could it be a Comfy UI limitation ?

---

To further improve the way this works in UE? widgets,

Maybe the text input pop up could be disabled if the "|" character is detected in the current string.
Because, in these cases, it's better to just edit the text directly in the widget, like we did before. 

![pop](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/ae254688-fd94-40a4-990e-babadb004436)

![mln](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/c96b8e1b-2f5e-4ee5-9b6d-1cd713a93a62)

**Comment by chrisgoringe** [comment]
> OK, the current autocomplete widget does have an advantage.
> 
> It can display a larger number of characters.   All single line string nodes, from different suites, seem to have the same limit on the characters displayed. Could it be a Comfy UI limitation ?

Yes. 

The single-line text widgets on ComfyUI are built-in (not by Comfy, but from the underlying LiteGraph code) with a fixed width. To add the autocomplete to these directly is not at all simple.

So going to a custom widget is basically unavoidable. And that (not the use of the html autocomplete) is what causes some of the UI complications. So swapping out to use a different autocomplete would add complexity without removing any. The other autocompletes knocking around in the custom node world are (AFAIK) working in the multi-line text area, which is a custom widget added by Comfy. And I really don't want to use a multi-line widget for the regexes!

Closing the autocomplete on detection of a specific character is fairly easy, but I wonder if it would be better to make the escape key do it (like is common in web UIs) but make it persistent (so if you press escape, the autoprompt goes away and doesn't come back!)

**Comment by JorgeR81** [comment]
> I really don't want to use a multi-line widget for the regexes

I agree. The text is not as visible in the multi line widget, so single line is better for me.

> if you press escape, the autoprompt goes away and doesn't come back

So the autocomplete would be disabled for that UE node ?

That would work well. 
For multi regex, we don't really need autocomplete.
